app/controllers/provider/sessions_controller.rb

```diff
@@ -74,47 +74,25 @@ def redirect_if_logged_in
   end
 
   def authenticate_user
-    captcha_is_available = request.post? # Internal strategy (user & pass)
-    return if captcha_is_available && !bot_check
+    strategy = Authentication::Strategy::InferService.call(auth_params, domain_account, @provider).result
 
-    params = if domain_account.settings.enforce_sso?
-               sso_params
-             else
-               request.post? ? auth_params : sso_params
-    end
-
-    # TODO: refactor the authentication flow.
-    # Right now, we have a hierarchy of classes, one for each auth strategy, and we manually instance the last child
-    # class, `ProviderOAuth2`, and then try all strategies one by one by calling `super` and going up in the hierarchy
-    # until a strategy works. Due to this, we can't know from the here which strategy was really used.
-    #
-    # The hierarchy right now is:
-    #
-    # `ProviderOAuth2` < `OAuth2Base` < `Internal` < `SSO` < `Base`
-    #
-    # This is very weird because `Internal`, which means "User + pass" is not a kind of `SSO`; and `OAuth2` is not
-    # a kind of `Internal`. Not to mention we are calling `SSO` to something which is merely a token authentication
-    # not related at all with any SSO.
-    strategy = Authentication::Strategy.build_provider(@provider)
-    user = strategy.authenticate(params)
+    return if strategy.bot_protected? && !bot_check
 
+    user = strategy.authenticate(auth_params)
     [user, strategy]
   end
 
   def auth_params
-    params.slice(:username, :password)
-  end
-
-  def sso_params
-    params.permit(%i[token expires_at redirect_url system_name code]).merge(request: request)
+    params.permit(*%i[username password ticket token expires_at redirect_url system_name code]).merge(request:)
```

Comment:
do you know what is this `merge(request:)` supposed to do?

Reply:
OAuth authentication strategies need it. They take some info from the request:

https://github.com/3scale/porta/blob/a27c8f084bb8941d99da903dddb569ea3e7d8997/app/lib/three_scale/oauth2/keycloak_client.rb#L40-L44